### ISSUE TITLE
Heredoc highlight after method definition

### ISSUE THREAD

**Issue by dsferreira** [vim-ruby/vim-ruby#319]
Using: `stable-20160928` tag

When using a heredoc after the method definition I'm seeing different behaviours with the highlight of the first heredoc tag.
Ex:

`EOS` start and end tags highlight the same colour:

```
def foo 
  <<-EOS
  EOS
end 
```

`EOS` start and end tags highlight in a different colour:

```
def foo(bar) 
  <<-EOS
  EOS
end 
```

**Comment by AndrewRadev** [comment]
The problem seems to be related to the negative lookbehind here: https://github.com/vim-ruby/vim-ruby/blob/e865f7cacf05908668ed07be2db553f4fb4d56fd/syntax/ruby.vim#L278-L281

From what I could debug, the `<<-EOS` is not considered a heredoc-starter, if it's preceded by `class`, or by a closing `)`, `]`, etc, or by `::`. Because there's a `\_s*` at the end of that pattern, this happens in the multiline case as well. So, both of these will highlight the `EOS` as a constant:

``` ruby
class <<EOS
end

class
  <<EOS
end
```

Which, according to ruby, seems to be correct (I guess this is the `class << self` pattern, don't know if it can be used for something else). However, in this case:

``` ruby
foo() <<EOS
foo()
  <<EOS
```
The first example is syntactically valid -- the result of running `foo()` is left-shifted by a `EOS` amount. The second example, is a syntax error (on ruby 2.3.1), ruby considers it an unclosed heredoc.

I created a branch, [fix-issue-319](https://github.com/vim-ruby/vim-ruby/tree/fix-issue-319), where I've tried to fix the issue. Oddly enough, the other heredoc patterns in the syntax file don't have this problem, so I copied them over. 

I have no idea if this is right, or what side effects it might have (which is why I made it into a branch and I'd appreciate you test it out for a while, @dsferreira). I'm quite sure I'd like for @dkearns to take a look, he might know why the difference in patterns and/or what side effects we might expect.
